### DISCUSSION TITLE
Chaining Node2Vec and GraphSAGE

### DISCUSSION THREAD

**Discussion by CalmLogarithm** [stellargraph/stellargraph#1873] [Q&A]
Eventually I will run this experiment. But while I find the time, I wanted to find out if it had been attempted before.

My end goal is inductive node embeddings via GraphSAGE. I want the inductive property so I can generate new embeddings without retraining the entire model.

My graph is an ontology. Generally speaking, there are two types of links: Taxonomic and semantic. Taxonomic links define the skeleton or structure of the ontology. Every node has an "IS-A" link such that there is a path between all nodes in a tree-like structure. The semantic links are the interesting pieces of information that connect nodes (social relationships, part-whole, etc.)

Currently I train it just in GraphSAGE and each node has a binary feature vector of length 12. My intuition tells me this isn't enough information for quality embeddings. **So the experiment is**: First, train embeddings using _just the taxonomic links fed into Node2Vec_. Second, use the embeddings from Node2Vec as the feature vectors for training embeddings using _just the semantic links fed into GraphSAGE_. Logically this makes sense to me as Node2Vec learns structures like what the taxonomic links describe. And GraphSAGE clusters related nodes using semantic information. Mathematically I don't know if this gains me anything with the added complexity...

**Comment by geoHeil** [comment]
I want to ask a similar question - perhaps even more generalized: https://github.com/stellargraph/stellargraph/discussions/1877

**Comment by adocherty** [comment]
Please let me know if my understanding is correct here. You'd like to:
* Take a knowledge graph with a single kind of link (taxonomic links)
* Use Node2Vec on this graph to give unsupervised embeddings for the nodes
* Use these embeddings as node features in a second graph constructed from the semantic links between the same nodes
* Use this graph to train GraphSAGE in supervised by node labels (I'm not sure what these node labels are)
* For ontological graphs you may gain some utility from knowledge graph methods.
* You don't give any information on what the binary features you are using for the node are, but adding more information from graph structure could be useful (as suggested in #1877)

Some notes for this approach:
* Using [Metapath2Vec](https://stellargraph.readthedocs.io/en/stable/demos/embeddings/metapath2vec-embeddings.html) will allow you to create embeddings using multiple link types (metapaths are random walks that respect link type see here
* You mention using GraphSAGE currently, but you have two types of links so using HinSAGE that can utilise different link and node types could help ([this demo](https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/hinsage-link-prediction.html) is for link prediction but HinSAGE can be used for node prediction and embeddings similar to GraphSAGE). Also look at [RGCN](https://stellargraph.readthedocs.io/en/stable/demos/node-classification/rgcn-node-classification.html).

An important point here is that both Node2Vec or MetaPath2Vec are NOT inductive. Any embeddings trained on one graph will not transfer to another graph, and furthermore for a graph that is disconnected (ie. is not a single connected component) the embeddings in one component will not have any relation to the embeddings for a disconnected component.

I hope this helps, please ask any follow up questions on this and some more information on how you are training GraphSAGE (do you use both types of links, are you training for link prediction or node classification?) and how you envisage the inductive training working would be useful.

**Reply by CalmLogarithm** [reply]
Essentially your understanding of my goals is correct.

I was vaguely aware of the HinSAGE/GraphSAGE distinction. That may be an easier experiment to try first. The next logical step would be training link prediction models for each of the two different link types.

The inductive property is key to the final work flow that I'm attempting:
1. Learn vector for out-of-vocabulary word.
2. Use vector and link prediction models to add a new node and edges to the graph.
3. Run the new node through the inductive model to generate a corresponding embedding (without retraining the model).

This would be an iterative, batch process. Eventually I would want to retrain the GraphSAGE/HinSAGE model to include the new words. But ideally there wouldn't be that many new words to learn so it could be infrequent.

If you're interested then here are the features I'm currently using. They are specific to the ontology model I'm using so they may not be very meaningful. That's why I didn't include them at first:

- Linguistic - Either language-specific or language-independent
- Instantiated - A concrete instance (like a proper noun) or an abstract one (common noun)
- Instantiable - Can be instantiated
- Transformable - Can be transformed into a different structure
- Semantic - Contains semantic information
- Syntactic - Contains syntactic information
- Morphology - Is a morphological derivation of a base form
- Orthography - Is an orthographic variant of a more common form

**Reply by CalmLogarithm** [reply]
I found a better explanation of my motivation. It comes from the GraphSAGE paper itself:

> While we focus on feature-richgraphs (e.g., citation data with text attributes, biological data with functional/molecular markers), ourapproach can also make use of structural features that are present in all graphs (e.g., node degrees).Thus, our algorithm can also be applied to graphs without node features.

My goal in using Node2Vec embeddings would be as structural features (a la node degrees as suggested in the paper). My intuition is that Node2Vec embeddings would provide richer structural features than node degrees along. It's important to my approach that I'm able to extend this structure with each new node. Given a new node, it would have one structural link that connects it to the existing graph.

**Comment by A2Amir** [comment]
Can Node2vec or other Embedding algorithms like Struc2Vec operate on completely disconnected nodes from the graph?

like the node below>
A----------->B------------>C   &nbsp; D(completely disconnected node from the graph)

**Reply by CalmLogarithm** [reply]
Yes, it can work. Which means that it will run. But I haven't evaluated the quality/utility of embedding disconnected nodes. Anecdotally, they seem to lie away from the other nodes.
